### ISSUE TITLE
support for fake clock (`new Date()`, `Date.now()`...) in vitest timers

### ISSUE THREAD

**Issue by cexbrayat** [vitest-dev/vitest#506]
### Clear and concise description of the problem

Jest support for timers has builtin support for clocks: when using `jest.useFakeTimers()`, a fake clock is instantiated. Then calls to `jest.advanceTimersByTime()` increase the fake clock time. Application code that use `new Date()`, or `Date.now()`, or `performance.now()` and others will get this fake time.

Jest leverages `@sinonjs/fake-timers` to do so (in its `modern` implementation, which is the default nowadays).

It would be super handy to have vitest do the same. Currently the API mimicks the one from Jest (which is awesome, don't get me wrong :hugs:) but doesn't handle this feature, which makes a test suite migration a bit more difficult, as tests that rely on `setTimeout` or`setInterval` will behave the same in vitest and Jest, but the one that leverage `Date` will not.

### Suggested solution

It could be possible to do the same as Jest and directly leverage `@sinonjs/fake-timers`.

### Alternative

It is possible for a developer who wants to have a fake clock in its test to use `@sinonjs/fake-timers` in addition to vitest

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by sheremet-va** [comment]
We can just mock date when interval moves in our implementation? We already have `vi.mockCurrentDate` that can do that.

I think, `@sinonjs/fake-timers` is too large for our needs

**Comment by sheremet-va** [comment]
But to be clear I am not opposing using it. I think if community decides, we can easily migrate to it. Maybe start a discussion in discord?

**Comment by cexbrayat** [comment]
Maybe just moving the mocked date could be enough. Ideally, these kinds of test would work out of the box:

```ts
test('new Date()', async() => {
  vi.useFakeTimers()

  const start = new Date().getTime()

  vi.advanceTimersByTime(2000)

  expect(new Date().getTime()).toBe(start + 2000)

  vi.useRealTimers()
})

test('Date.now()', async() => {
  vi.useFakeTimers()

  const start = Date.now()

  vi.advanceTimersByTime(2000)

  expect(Date.now()).toBe(start + 2000)

  vi.useRealTimers()
})
```

**Comment by fadi-george** [comment]
I feel `mockCurrentDate` should be changed (back?) to `setSystemTime` to make it more similar/consistent to jest.

**Comment by Smrtnyk** [comment]
its not just moving the current time, its also about scheduling timers and firing them when they should fire.
Maybe it will need to wrap for example setTimeout and keep track of callbacks and timeouts and fire them when time is moved.
Probably sinon fake timers is the least painful approach and is already battletested

**Comment by sheremet-va** [comment]
> its not just moving the current time, its also about scheduling timers and firing them when they should fire.

It already does it tho?

**Comment by sheremet-va** [comment]
I must've mistaken sinon/fake-timers with sinon. Now that I look at it, I think we can implement jest like methods using sinon/fake-timers inside without gaining much weight. Will make a PR soon.

**Comment by sheremet-va** [comment]
Released in [v0.2.0](https://github.com/vitest-dev/vitest/releases/tag/v0.2.0)

**Comment by patak-dev** [comment]
@RPDeshaies you would call [`vi.useRealTimers()`](https://vitest.dev/api/#vi-userealtimers). See the example here https://vitest.dev/api/#vi-setsystemtime
```js
const date = new Date(1998, 11, 19)

vi.useFakeTimers()
vi.setSystemTime(date)

expect(Date.now()).toBe(date.valueOf())

vi.useRealTimers()
```
